File: python/731.py
"""Project Euler Problem 731: A Stoneham Number.

Starting at the Nth digit after the decimal point of the decimal expansion of
Σ_{i=1}^∞ 1/(3^i 10^(3^i)), find the next K digits.

For each i, we need to find the K digits of 1 / 3^i starting from digit
N - 3^i. This is just the last K digits of 10^{N+K-3^i} / 3^i, but we need
to subtract the fractional part, which is 10^{N+K-3^i} (mod 3^i) / 3^i. To
protect against carries from the next few digits, we fetch K+2 digits instead
of K, and only take the first K digits at the end.
"""

from __future__ import annotations


def pow_mod(base: int, exp: int, mod: int) -> int:
    """Fast exponentiation modulo mod."""
    result = 1
    base = base % mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result


def mod_inv(a: int, m: int) -> int:
    """Modular inverse using extended Euclidean algorithm."""
    if m == 1:
        return 0
    t, new_t = 0, 1
    r, new_r = m, a % m
    while new_r != 0:
        q = r // new_r
        t, new_t = new_t, t - q * new_t
        r, new_r = new_r, r - q * new_r
    if r != 1:
        raise ValueError("Modular inverse does not exist")
    if t < 0:
        t += m
    return t


def solve() -> str:
    """Solve Problem 731."""
    n = 10**16
    k = 10
    t = 3
    b = 10
    m = b ** (k + 2)

    ans = 0
    i = 1
    while t**i < n:
        three_power_i = t**i
        exp = n + k + 1 - three_power_i
        if exp > 0:
            term1 = pow_mod(b, exp, m)
            term2 = pow_mod(b, exp, three_power_i)
            inv_three_i = mod_inv(three_power_i, m)
            ans = (ans + (term1 - term2) * inv_three_i) % m
        i += 1

    return str(ans % m)[:k]


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return int(result)


if __name__ == "__main__":
    main()

File: c/731.c
/* Project Euler 731: A Stoneham Number.
 * Compute digits of Stoneham number sum 1/(3^i * 10^(3^i)).
 * Uses modular arithmetic with mod = 10^12 (K+2 digits).
 *
 * We need arbitrary precision modular exponentiation with very large moduli
 * (3^i can be huge, up to 3^34 ~ 10^16). We use __int128 for intermediate results.
 */
#include <stdio.h>
#include <stdint.h>

typedef long long ll;
typedef unsigned long long ull;
typedef __int128 lll;

/* We need modular exponentiation where mod can be up to ~10^16 */
ull pow_mod_ull(ull base, ull exp, ull mod) {
    if (mod == 1) return 0;
    ull result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp & 1) result = (unsigned __int128)result * base % mod;
        base = (unsigned __int128)base * base % mod;
        exp >>= 1;
    }
    return result;
}

/* Extended GCD */
ll ext_gcd(ll a, ll b, ll *x, ll *y) {
    if (a == 0) { *x = 0; *y = 1; return b; }
    ll x1, y1;
    ll g = ext_gcd(b % a, a, &x1, &y1);
    *x = y1 - (b / a) * x1;
    *y = x1;
    return g;
}

ll mod_inv_ll(ll a, ll m) {
    ll x, y;
    ext_gcd(a % m, m, &x, &y);
    return ((x % m) + m) % m;
}

int main() {
    ull n = 10000000000000000ULL; /* 10^16 */
    int k = 10;
    ull m = 1000000000000ULL; /* 10^(k+2) = 10^12 */

    ull ans = 0;

    /* Compute 3^i iteratively and check 3^i < n */
    ull three_power = 3; /* 3^1 */
    int i = 1;
    while (three_power < n) {
        ull exp = n + k + 1 - three_power;
        /* term = (10^exp mod m - 10^exp mod 3^i) * inv(3^i, m) mod m */
        /* But we need 3^i for modular ops */
        ull term1 = pow_mod_ull(10, exp, m);
        ull term2 = pow_mod_ull(10, exp, three_power);

        /* inv_three_i = modular inverse of three_power mod m */
        ll inv_three_i = mod_inv_ll((ll)(three_power % m), (ll)m);

        /* (term1 - term2) might be negative in unsigned, handle carefully */
        ll diff = (ll)term1 - (ll)term2;
        ll contribution = (lll)diff * inv_three_i % (ll)m;
        contribution = ((contribution % (ll)m) + (ll)m) % (ll)m;

        ans = (ans + (ull)contribution) % m;

        /* Next 3^i -- check for overflow */
        if (three_power > n / 3) break; /* would overflow */
        three_power *= 3;
        i++;
    }

    /* Extract first k digits from ans (which has k+2 digits) */
    /* ans is mod 10^12, we want first 10 digits = ans / 100 */
    ull result = ans / 100; /* divide by 10^2 to get first k digits */

    printf("%llu\n", result);
    return 0;
}
